staging/src/k8s.io/client-go/tools/cache/listwatch.go

```diff
@@ -171,6 +180,24 @@ type ListWatch struct {
 
 	// DisableChunking requests no chunking for this list watcher.
 	DisableChunking bool
+
+	// WatchListSemanticsSupported indicates whether a client explicitly supports
+	// WatchList semantics.
+	//
+	// Over the years, unit tests in kube have been written in many different ways.
+	// After enabling the WatchListClient feature by default, existing tests started failing.
+	// To avoid breaking existing client-go users after upgrade,
+	// we introduced this field as an opt-in.
+	//
+	// If a client implements the IsWatchListSemanticsSupported method, this field
+	// is set to true and the reflector will use WatchList if the feature was enabled.
+	// If the method is not defined, the reflector will automatically disable
+	// the WatchList feature.
+	//
+	// This method has been added to all standard clients except for the FakeClient.
+	// Adding support for FakeClient was deliberately avoided, since many existing
+	// unit tests rely on its legacy behavior and would otherwise fail in subtle ways.
+	WatchListSemanticsSupported bool
```

Comment:
I'm not convinced about this part.

While I'm fine with FakeClient to not support watch-list, I actually would like to have every implementation that can potentially be used in production code to support it...

Reply:
An alternative would be to explicit disable the feature for certain clients. Some time ago we opened a [POC PR](https://github.com/kubernetes/kubernetes/pull/134180) with that approach (The feature was disabled for fake clients). Unfortunately with this approach some tests will still fail and would require fixes like https://github.com/kubernetes/kubernetes/pull/134180/commits/972479430596113c5f03c0c806aff7802a4754d1